### ISSUE TITLE
[4.0] Attempt select CUSTOM Module OPTION Background image results in Internal Server Error

### ISSUE THREAD

**Issue by sdrawdenek** [joomla/joomla-cms#30045]
### Steps to reproduce the issue

Open Custom Module screen.

Select OPTIONS tab.

Browse Images to select an image (SELECT) - image 1.

### Expected result

Expect the image to be selectable - it was in J4 Beta 2 for Testing.

### Actual result

Internal Server Error - image 2.

### System information (as much as possible)

PHP 7.3

J4 Dev 3

### Additional comments

Not an issue on my server when I use J4 Beta 2 for Testing![screen shot 2020-07-07 at 18 00 08](https://issues.joomla.org/uploads/1/83353dda8fc1ddb88d4839e5ca9c0891.png)![screen shot 2020-07-07 at 18 00 09](https://issues.joomla.org/uploads/1/cc603cc4bc92c5133dca7f828a91a756.png)

**Comment by Quy** [comment]
Unable to reproduce. Please try testing with the nightly builds. https://developer.joomla.org/nightly-builds.html

**Comment by PhilETaylor** [comment]
Are you selected an image? because without selecting an image you get https://github.com/joomla/joomla-cms/issues/29497

**Comment by sdrawdenek** [comment]
Yes I am selecting an image BEFORE then selecting SELECT button.

And I have just tried again after installing the latest lightly build.

Still see Internal Server Error.

**Comment by PhilETaylor** [comment]
ok I can replicate this (In Firefox, [as impossible in Safari due to there being no select button](https://github.com/joomla/joomla-cms/issues/29314)), but with a very specific set of clicks. 

If you click the IMAGE then it works fine.

If you SELECT the checkbox in the top left of the image and then 

Nothing is logged as a Ajax request and nothing in the console.log but there is a reproducible problem here.

**Comment by PhilETaylor** [comment]
ok I can replicate this (In Firefox, [as impossible in Safari due to there being no select button](https://github.com/joomla/joomla-cms/issues/29314)), but with a very specific set of clicks. 

If you click the IMAGE then it works fine.

If you DONT CLICK THE IMAGE - but instead SELECT the checkbox in the top left of the image and then click the SELECT button in the modal footer then you replicate the problem. 

Nothing is logged as a Ajax request and nothing in the console.log but there is a reproducible problem here.

**Comment by Quy** [comment]
@dgrammatiko Please see https://github.com/joomla/joomla-cms/issues/30045#issuecomment-656058764. Thanks.

**Comment by dgrammatiko** [comment]
@Quy there you go:

For issue #29314 Someone added an `overflow: hidden` for the class `.input-group-append`, removing it the field is rendered correctly:
![Screenshot 2020-07-10 at 11 59 25](https://user-images.githubusercontent.com/3889375/87142801-57674700-c2a5-11ea-8711-25bfc2682eb7.png)

> If you DONT CLICK THE IMAGE - but instead SELECT the checkbox in the top left of the image and then click the SELECT button in the modal footer then you replicate the problem.

I cannot reproduce the error  reported in #29497 

![Jul-10-2020 12-09-58](https://user-images.githubusercontent.com/3889375/87143436-5c78c600-c2a6-11ea-8dda-f2aacee08705.gif)

Finally the Issue reported here cannot be confirmed as I'm getting a PHP `class Folder not found`

**Comment by PhilETaylor** [comment]
Im sorry, but today with a0116ab6bbf52c9772dbfedc91edcf9a49649146 and Safari Version 13.1.2 (15609.3.5.1.3) I can still replicate this issue reported

Also I can still replicate https://github.com/joomla/joomla-cms/issues/29497 as well.

This is a clean install of 4.0-dev from git with `jinstall` asset build on my laptop after a reformat of the hard drive!

**Comment by dgrammatiko** [comment]
> Also I can still replicate #29497 as well.

What user? Eg what are the permissions for the testing user

Also what is your configuration: subfolder/root install, folder/files permissions, etc

I cannot help if I cannot replicate the issue...

**Comment by PhilETaylor** [comment]
Super User (The only user) created when installing.

In the root. So url is http://127.0.0.1/

in a docker container with 777 permissions on everything.

**Comment by dgrammatiko** [comment]
> In the root. So url is http://127.0.0.1/

Can you attach a fake domain (eg make an alias in the hosts file) so the Ajax request will be `http://something.local` as this is something I cannot test on my setup?

**Comment by PhilETaylor** [comment]
If you have 5 mins YOU can go to my local development here: https://97d052c4e6c2.ngrok.io/administrator/

u: admin
p: admin

I'll leave this tunnel up and grab a coffee - let me know if/when you are finished so I can close the tunnel.

**Comment by dgrammatiko** [comment]
@PhilETaylor I still cannot replicate it 
![Jul-10-2020 12-43-44](https://user-images.githubusercontent.com/3889375/87146500-51746480-c2ab-11ea-96fc-54b9a15cc6f9.gif)

Are you too fast on clicking the select button so there might be a race condition as the iframe needs to initialise its js and sync with the parent window?

**Comment by PhilETaylor** [comment]
Very strange. 

I thought at first it could be a race condition but Im bring deliberately slow (10-15 seconds between each click) and I still get `Internal Server Error` when either 1) Not selecting an image and clicking select to close modal and 2) selecting the checkbox of an image and clicking select to close the modal. 

I DONT GET AN ERROR when I click the actual image (as opposed to clicking the checkbox in the top left corner) so there is something different between clicking the actual IMAGE and clicking the checkbox.

**Comment by PhilETaylor** [comment]
same in Firefox so not a safari beta problem :(

**Comment by PhilETaylor** [comment]
Just asked "a bog standard user" in a different country (@redevo) to try the same thing using the link I gave you - he could replicate it :( 

![h6czel96](https://user-images.githubusercontent.com/400092/87147336-5eda2080-c2a4-11ea-9ec2-a9b77f0cc32c.png)

**Comment by dgrammatiko** [comment]
> so there is something different between clicking the actual IMAGE and clicking the checkbox.

There IS. The images to allow double click have some weird code (I think it's a vue dependency) but what's funny is that we're testing using the same safari...

Ok one last idea: do you have .htaccess enabled? if so can you disable it, clear the browser cache and retry (you might be running an older version of the js script, the .gz)

**Comment by PhilETaylor** [comment]
Its nginx - so no .htaccess :)

**Comment by RedEvo** [comment]
nginx has form….

:)


David Robinson
david@redevolution.com
Mobile: 07710 461776
Office: 01224 443551

**Comment by PhilETaylor** [comment]
oh hang on - just seen this in the log:

`GET /index.php/component/ajax                                       200 OK`

is something routing to the frontend instead of the backend?

that's not the keepalive as that has correct /administrator/:

```
  "system.keepalive": {
            "interval": 840000,
            "uri": "\/administrator\/index.php?option=com_ajax&amp;format=json"
        },
```

**Comment by dgrammatiko** [comment]
@PhilETaylor I'm out of ideas, the only proper way to figure out there is to place a breakpoint in the com_media controller and inspect what's the input. Might be a problem with csrf or some other data send back to the server

**Comment by PhilETaylor** [comment]
I have an error - one mo...

**Comment by PhilETaylor** [comment]
go here https://97d052c4e6c2.ngrok.io/administrator/index.php?option=com_content&view=article&layout=edit

click select 

double click sample data 

<img width="1600" alt="Screenshot 2020-07-10 at 12 04 44" src="https://user-images.githubusercontent.com/400092/87148032-8ed5f380-c2a5-11ea-8a15-a6ca974e5a7b.png">

no idea if that's connected :(

**Comment by PhilETaylor** [comment]
> Might be a problem with csrf or some other data send back to the server

Im not seeing any network traffic or Ajax calls :(

**Comment by dgrammatiko** [comment]
> Im not seeing any network traffic or Ajax calls :(

OK, we're getting somewhere, Can you attach a fake domain (eg make an alias in the hosts file) as I suggested before. I know it sounds weird but...

**Comment by PhilETaylor** [comment]
Also when I click the checkbox in the top left of an image - the BACKGROUND (the main article edit page, not the modal) moves - like a JS warning/error is being inserted and removed. 

The ngrok is a fake domain, when you use that Joomla believes that's the domain its on and all Ajax calls go back to that domain.

**Comment by dgrammatiko** [comment]
I'm guessing you're using docker and I'm using a local server (no vm). There might be some firewall or something else that blocks XHR requests? Nginx probably is not the problem here as I already have a perfectly working instance (sloth-theme.site)

**Comment by PhilETaylor** [comment]
one mo and I'll move to `php -S 0.0.0.0:80` outside of docker and outside of nginx

**Comment by PhilETaylor** [comment]
made no difference - same issues. 

So this rules out docker. Rules out nginx.

:-(

**Comment by dgrammatiko** [comment]
Ok one more navigate to the page and open the media field modal. Open the inspector and go to the console. On the bottom right there should be a select option, select `Page`, then type `Joomla` and return (on the left side, the terminal lookalike)
You should have an empty object named `selectedFile` :
![Screenshot 2020-07-10 at 13 25 34](https://user-images.githubusercontent.com/3889375/87149803-3147a400-c2b1-11ea-8f16-c0a541b7435b.png)

**Comment by PhilETaylor** [comment]
Correct. 

When I click the checkbox Its still an empty object

When I click the actual image Its a populated object

**Comment by dgrammatiko** [comment]
> When I click the checkbox Its still an empty object

Out of curiosity what's your options on your clicking device?
![Screenshot 2020-07-10 at 13 33 28](https://user-images.githubusercontent.com/3889375/87150264-0447c100-c2b2-11ea-9623-92310d3f2895.png)

**Comment by PhilETaylor** [comment]
haha - its a Magic Mouse (that needs charging :) )

<img width="673" alt="Screenshot 2020-07-10 at 12 34 46" src="https://user-images.githubusercontent.com/400092/87150348-c5157200-c2a9-11ea-9586-6a41120a4a92.png">

**Comment by PhilETaylor** [comment]
want another play? https://dd1ca95bb272.ngrok.io

**Comment by dgrammatiko** [comment]
Just to narrow this issue, can you edit: 
https://github.com/joomla/joomla-cms/blob/a0116ab6bbf52c9772dbfedc91edcf9a49649146/administrator/components/com_media/resources/scripts/components/browser/items/image.vue#L2
and remove `@dblclick="openPreview()" @mouseleave="hideActions()"` from that line

also  change `@click.stop="toggleSelect()"` to `@click="toggleSelect()"` on https://github.com/joomla/joomla-cms/blob/a0116ab6bbf52c9772dbfedc91edcf9a49649146/administrator/components/com_media/resources/scripts/components/browser/items/image.vue#L12

Then recompile with `npm install `

**Comment by PhilETaylor** [comment]
**Done - that FIXES the selection issues :)** 

It doesnt fix the issue where you click select button to close the modal without ever attempting to select anything.

**Comment by PhilETaylor** [comment]
it doesnt fix double clicking a FOLDER - that still gives a console error

**Comment by dgrammatiko** [comment]
> Done - that FIXES the selection issues

It's not an actual fix, we just removed some listeners to narrow down the problem, which turn out to be a race condition between mouse enter/ mouse click

> It doesnt fix the issue where you click select button to close the modal without ever attempting to select anything.

That's weird because you've confirmed before that there is an empty object there and the code should handle this correctly 😞

**Comment by PhilETaylor** [comment]
well at least this is some kind of progress :)

**Comment by dgrammatiko** [comment]
@PhilETaylor can you test https://github.com/joomla/joomla-cms/pull/30066

It solves the error message when clicking on select without any file selected. The other part on the image clicking misbehaving needs a bit more debugging (also I'm not very familiar with vue, so someone else can fix it faster)

**Comment by sdrawdenek** [comment]
Hi, hope this helps you guys.

I was using default installation of J4 Dev 3 when I reported - and TinyMCE was default content editor.

Today after installing JCE (latest release) I retested Custom Module Background Image. It worked. No Internal Server Error.

When I changed default content editor to any one of NONE, CODEMIRROR, TINYMCE I saw the Internal Server Error message when I tried to add a background image.

Best wishes

**Comment by PhilETaylor** [comment]
IIRC JCE replaces the media browser with one of its own

**Comment by dgrammatiko** [comment]
@sdrawdenek please test #30066

**Comment by PhilETaylor** [comment]
#30066 fixes `the issue where you click select button to close the modal without ever attempting to select anything. which is Issue https://github.com/joomla/joomla-cms/issues/29497`

#30066 fixes `Error message shown when clicking the checkbox and then select button to close modal`

#30066 doesnt fix `When using the checkbox to select an image, and then clicking the select button to close model, no image is actually selected in the page`

#30066 doesnt fix `when double clicking on a folder in media manager I get TypeError: undefined is not an object (evaluating 'this.$refs.actionToggle.focus')`

**Comment by dgrammatiko** [comment]
The last 2 are definitely associated with the media manager app (the vue js code)

**Comment by PhilETaylor** [comment]
I guess, for clarity, we should ask @sdrawdenek to close this issue (as its resolved by #30066), and open a new issue for the two remaining problems? 

Also can I say a big thank you - without your help today these issues would have dragged on, we needed an expert :) and I stick to jQuery ;-) ;-)
